simplicity-sys/src/dag.rs

```diff
@@ -0,0 +1,152 @@
+// Simplicity Bindings
+// Written in 2022 by
+//   Christian Lewe <clewe@blockstream.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+use crate::bitstream::Bitstream;
+use crate::bitstring::Bitstring;
+use crate::error::Error;
+use crate::util;
+use libc::{c_void, size_t};
+use std::fmt;
+
+/// Kind of Simplicity node.
+#[repr(C)]
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum Tag {
+    COMP,
+    CASE,
+    ASSERTL,
+    ASSERTR,
+    PAIR,
+    DISCONNECT,
+    INJL,
+    INJR,
+    TAKE,
+    DROP,
+    IDEN,
+    UNIT,
+    HIDDEN,
+    WITNESS,
+    JET,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+union AuxTypes {
+    /// scratch space for verifyCanonicalOrder
+    aux: size_t,
+    /// source type, target type
+    types: [size_t; 2],
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+union IndicesChildrenWitness {
+    /// source index, target index
+    indices: [size_t; 2],
+    /// first child, second child
+    children: [size_t; 2],
+    /// witness
+    witness: Bitstring,
+}
+
+#[repr(C)]
+struct PrivateDagNode {
+    jet: *const c_void,
+    cmr: [u32; 8],
+    aux_types: AuxTypes,
+    indices_children_witness: IndicesChildrenWitness,
+    tag: Tag,
+}
+
+extern "C" {
+    fn decodeMallocDag(
+        dag: *mut *mut PrivateDagNode,
+        combinator_counters: *const c_void,
+        stream: *mut Bitstream,
+    ) -> i32;
+}
+
+/// Simplicity DAG (typed or untyped).
+/// Uses the Elements extension by default.
+pub struct Dag {
+    first: *mut PrivateDagNode,
+    len: usize,
+}
+
+impl<'a> fmt::Debug for Dag {
```

Comment:
In a0faaf2957e71fc5e72a23d41e916703e739bb53:

nit: this `'a` shouldn't be there. I'm not sure why the compiler doesn't warn about this.